File: SimplePlus/Core/Serializer.cpp
#include "Serializer.h"

const int Serializer::ReadInt(const char* data, int* readCtx) {
    int val = *((int*)&data[*readCtx]);
    *readCtx += sizeof(int);
    return val;
}

const float Serializer::ReadFloat(const char* data, int* readCtx) {
    int val = *((float*)&data[*readCtx]);
    *readCtx += sizeof(float);
    return val;
}

const double Serializer::ReadDouble(const char* data, int* readCtx) {
    double val = *((double*)&data[*readCtx]);
    *readCtx += sizeof(double);
    return val;
}

const char* Serializer::ReadString(const char* data, int* readCtx) {
    int len = ReadInt(data, readCtx);

    if (len <= 0 && len > 128) {
        printf("[Serializer]: reading string lenght was outside of bonds: %d", len);
        return NULL;
    }

    const char* value = &data[*readCtx];
    *readCtx += len;
    return value;
}

void Serializer::WriteInt(int value, const char* data, int* writeCtx) {
    *((int*)&data[*writeCtx]) = value;
    *writeCtx += sizeof(int);
}

void Serializer::WriteFloat(float value, const char* data, int* writeCtx) {
    *((float*)&data[*writeCtx]) = value;
    *writeCtx += sizeof(float);
}

void Serializer::WriteDouble(double value, const char* data, int* writeCtx) {
    *((double*)&data[*writeCtx]) = value;
    *writeCtx += sizeof(double);
}

void Serializer::WriteString(const char* value, char* data, int* writeCtx) {
    int len = strlen(value);

    if (len <= 0 && len > 128) {
        printf("[Serializer]: writing string lenght was outside of bonds: %d", len);
        return;
    }

    WriteInt(len, data, writeCtx);
    memcpy(&data[*writeCtx], value, len);
    *writeCtx += len;

    return;
}

File: SimpleNet/Core/Serializer.cs

using System;
using System.Text;

namespace SimpleNET
{
    public class Serializer
    {
        public static int ReadInt(byte[] buffer, ref int readIndex)
        {
            var value = BitConverter.ToInt32(buffer, readIndex);
            readIndex += sizeof(int);
            return value;
        }

        public static float ReadFloat(byte[] buffer, ref int readIndex)
        {
            var value = BitConverter.ToSingle(buffer, readIndex);
            readIndex += sizeof(float);
            return value;
        }

        public static double ReadDouble(byte[] buffer, ref int readIndex)
        {
            var value = BitConverter.ToDouble(buffer, readIndex);
            readIndex += sizeof(double);
            return value;
        }

        public static string ReadString(byte[] buffer, ref int readIndex)
        {
            var len = ReadInt(buffer, ref readIndex);
            var value = Encoding.ASCII.GetString(buffer, readIndex, len);
            readIndex += value.Length;
            return value;
        }

        public static void WriteInt(int value, ref byte[] buffer, ref int writeIndex)
        {
            var valBuf = BitConverter.GetBytes(value);
            Buffer.BlockCopy(valBuf, 0, buffer, writeIndex, sizeof(int));
            writeIndex += sizeof(int);
        }

        public static void WriteFloat(float value, ref byte[] buffer, ref int writeIndex)
        {
            var valBuf = BitConverter.GetBytes(value);
            Buffer.BlockCopy(valBuf, 0, buffer, writeIndex, sizeof(float));
            writeIndex += sizeof(float);
        }

        public static void WriteDouble(double value, ref byte[] buffer, ref int writeIndex)
        {
            var valBuf = BitConverter.GetBytes(value);
            Buffer.BlockCopy(valBuf, 0, buffer, writeIndex, sizeof(float));
            writeIndex += sizeof(float);
        }

        public static void WriteString(string value, ref byte[] buffer, ref int writeIndex)
        {
            WriteInt(value.Length, ref buffer, ref writeIndex);
            var valBuf = Encoding.ASCII.GetBytes(value);
            Buffer.BlockCopy(valBuf, 0, buffer, writeIndex, valBuf.Length);
            writeIndex += valBuf.Length;
        }
    }
}
